src/hooks/useImageUploader.ts

```diff
@@ -0,0 +1,111 @@
+import type { ChangeEventHandler, MouseEventHandler } from 'react'
+import type {
+  ImageInfo,
+  ImageUploaderProps,
+  UploaderProps
+} from '@components/ImageUploader'
+import { useRef, useState } from 'react'
+import { v4 as uuidV4 } from 'uuid'
+
+const isValidImageUrl = (file: unknown): file is string =>
+  typeof file === 'string'
+
+const MAX_LIST_LENGTH = 10
+const NOTICE_MESSAGE = '사진은 최대 10장만 추가 가능합니다.'
+
+export const useImageUploader = ({
+  imageList: defaultImageList,
+  onChange
+}: ImageUploaderProps): UploaderProps => {
+  const uploaderRef = useRef<HTMLInputElement | null>(null)
+  const imageListRef = useRef<HTMLDivElement | null>(null)
+  const [imageList, setImageList] = useState<ImageInfo[]>(defaultImageList)
+
+  const openUploader = (): void => {
+    if (imageList.length === 10) {
+      alert(NOTICE_MESSAGE)
+      return
+    }
+
+    uploaderRef.current?.click()
+  }
+
+  const removeImage: MouseEventHandler<HTMLDivElement> = e => {
+    if (!imageListRef.current) {
+      return
+    }
+
+    const closeIcons = imageListRef.current.querySelectorAll(
+      '[data-id="close-icon"]'
+    )
+    const fileIndex = Array.from(closeIcons).findIndex(
+      icon => icon === e.target
+    )
+
+    const firstToIndexFileList = imageList.slice(0, fileIndex)
+    const indexToLastFileList = imageList.slice(fileIndex + 1)
+    const newFiles = [...firstToIndexFileList, ...indexToLastFileList]
+
+    const isRepresent = imageList[fileIndex].isRepresent && imageList.length > 1
+    if (isRepresent) {
+      newFiles[0].isRepresent = true
+    }
+
+    onChange({ eventType: 'remove', imageList: newFiles })
+    setImageList(newFiles)
+  }
+
+  const addImage: ChangeEventHandler<HTMLInputElement> = async e => {
+    const files = e.target.files
+
+    if (!files) {
+      return
+    }
+
+    const isOverListLength =
+      MAX_LIST_LENGTH - (imageList.length + files.length) < 0
+    if (isOverListLength) {
+      alert(NOTICE_MESSAGE)
+      return
+    }
+
+    const fulfilledImageFiles = Array.from(files).map(file => {
+      const reader = new FileReader()
+
+      return new Promise(resolve => {
+        reader.onload = (): void => {
+          resolve(reader.result)
+        }
+
+        reader.readAsDataURL(file)
+      })
+    })
+
+    const imageFiles = await Promise.all(fulfilledImageFiles)
+    imageFiles.forEach((file, index) => {
+      const isRepresent = imageList.length === 0 && index === 0
+      const imageUrl = isValidImageUrl(file) ? file : ''
+      const newImage = {
+        id: uuidV4(),
+        isRepresent,
+        url: imageUrl
+      }
+
+      setImageList(prevImageList => {
+        return [...prevImageList, newImage]
+      })
+    })
+
+    onChange({ eventType: 'upload', imageList })
```

Comment:
map을 통해서 새로운 배열을 만들고 setImageList를 한번만 해주는 방식은 어떤가용 ? 
setState를 forEach문을 돌리면서 하나의 파일을 만들 때 마다 해주고 있어서 모든 파일을 수정한 후
한번에 setState를 해주는 방식이 더 좋을것 같다는 생각이 들어요 !


```suggestion
    const newImageList = imageFiles.map((file, index) => {
      const isRepresent = imageList.length === 0 && index === 0
      const imageUrl = isValidImageUrl(file) ? file : ''
      
      return {
        id: uuidV4(),
        isRepresent,
        url: imageUrl
      }
    })
    
    setImageList(newImageList)
    
    onChange({ eventType: 'upload', imageList: newImageList })
```

Reply:
요 부분 [ccd15126](https://github.com/price-offer/offer-design-system/commit/ccd15126bf7d200bfb57970b8ce3f7b8bb916599) 요기서 반영 했습니당! 감사합니당 ㅎㅎ